<a href="https://colab.research.google.com/github/senisekoni32-a11y/ml_internship_tasks/blob/main/Task_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [5]:
import pandas as pd

df = pd.read_csv("iris.csv")
print(df.shape)
print(df["species"].value_counts())
df.head()

(150, 5)
species
setosa        50
versicolor    50
virginica     50
Name: count, dtype: int64


,sepal_length,sepal_width,petal_length,petal_width,species
0,5.1,3.5,1.4,0.2,setosa
1,4.9,3.0,1.4,0.2,setosa
2,4.7,3.2,1.3,0.2,setosa
3,4.6,3.1,1.5,0.2,setosa
4,5.0,3.6,1.4,0.2,setosa


In [6]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

X = df.drop("species", axis=1)      # clues: the four measurements
y = df["species"]                   # answers: the species

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

print(X_train.shape, X_test.shape)

(120, 4) (30, 4)


In [10]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

knn = KNeighborsClassifier(n_neighbors=5)    # K = 5
knn.fit(X_train,y_train)                            # learn from the training clues and training answers
y_pred = knn.predict(X_test)                    # guess species for the test clues

print(accuracy_score(y_test, y_pred))

0.9333333333333333


In [13]:
from sklearn.metrics import confusion_matrix, classification_report

print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

[[10  0  0]
 [ 0  9  1]
 [ 0  2  8]]
              precision    recall  f1-score   support

      setosa       1.00      1.00      1.00        10
  versicolor       0.82      0.90      0.86        10
   virginica       0.89      0.80      0.84        10

    accuracy                           0.90        30
   macro avg       0.90      0.90      0.90        30
weighted avg       0.90      0.90      0.90        30



In [17]:
for k in [1, 3, 5, 7, 9, 15, 25]:
    knn = KNeighborsClassifier(n_neighbors=k)
    knn.fit(X_train, y_train)
    y_pred = knn.predict(X_test)
    print(k, accuracy_score(y_test, y_pred))

1 0.9666666666666667
3 0.9333333333333333
5 0.9333333333333333
7 0.9666666666666667
9 0.9666666666666667
15 0.9666666666666667
25 0.9


**Results**

This task was aimed at introducing myself to how K-Nearest Neighbour Models work: Classification based on characteristic similarities.

Here, the dataset used was one detailing the physical properties of an equal number of 3 flower species and the model did the work of classifying them into species.

For the first KNN model, I used k=5, and got a 93% accuracy. The 2 errors were virginica mistaken for versicolor, and this is likely because the two flower species have similar petal and sepal dimensions in the dataset.

In using a for-loop to check for other values of k (1, 3, 5, 7, 9, 15, 25), values 7, 9 and 15 performed best at 97% and k=25 performed worst at 90%.

k=1 also yielded 97%, but it lets a single neighbour decide each prediction, so its vote is all-or-nothing and one unusual flower can mislead it. I trust the stable range of 7 to 15 more. With only 30 test flowers, each flower is worth about 3 percentage points, so most of these differences are small.